mesonbuild/utils/universal.py

```diff
@@ -1126,6 +1127,26 @@ def determine_worker_count(varnames: T.Optional[T.List[str]] = None) -> int:
             num_workers = 1
     return num_workers
 
+def is_parent_path(parent: str, trial: str) -> bool:
+    '''Checks if @trial is a file under the directory @parent. Both @trial and @parent should be
+       adequately normalized, though empty and '.' segments in @parent and @trial are accepted
+       and discarded, matching the behavior of os.path.commonpath.  Either both or none should
+       be absolute.'''
+    assert os.path.isabs(parent) == os.path.isabs(trial)
+    if is_windows():
+        parent = parent.replace('\\', '/')
+        trial = trial.replace('\\', '/')
+
+    split_parent = parent.split('/')
+    split_trial = trial.split('/')
+
+    split_parent = [c for c in split_parent if c and c != '.']
+    split_trial = [c for c in split_trial if c and c != '.']
+
+    components = len(split_parent)
+    return len(split_trial) >= components and split_trial[:components] == split_parent
```

Comment:
The length check is not needed: two lists of different length cannot be equal and Python has the same short circuit built in. But I feel we are bikesheeding around an irrelevant issue.

Reply:
You still have the cost of `split_trial[:components]`, which is O(components) in Python (`list` has no copy-on-write or similar optimizations):

```
>>> timeit.timeit('x[:100]', globals={'x':list(range(10000))})
0.42416683500050567
>>> timeit.timeit('x[:1000]', globals={'x':list(range(10000))})
2.619175838000956
```

Reply:
I'm procrastinating something rather boring, thus I'll delve into the realm of micro-optimization 🙂: I don't think the average number of components in a path is that large, and you need to balance the cost of two function calls (which in Python are expensive) with the cost of list slicing.

Am I looking at it wrong?
```
>>> timeit.timeit('x == y[:20]', globals={'x': list(range(20)), 'y': list(random.randint(0, 20) for _ in range(100))})
0.1466136999661103
>>> timeit.timeit('len(y) >= 20 and x == y[:20]', globals={'x': list(range(20)), 'y': list(random.randint(0, 20) for _ in range(100))})
0.20071899995673448
```